pkg/controller/runtime/adapter.go

```diff
@@ -479,3 +499,52 @@ func (adapter *adapter) runOnce(ctx context.Context, logger *zap.Logger) (err er
 
 	return adapter.ctrl.Run(ctx, adapter, logger)
 }
+
+func outputTrackingID(md *resource.Metadata) string {
+	return fmt.Sprintf("%s/%s/%s", md.Namespace(), md.Type(), md.ID())
+}
+
+func (adapter *adapter) StartTrackingOutputs() {
+	if adapter.outputTracker != nil {
+		panic("output tracking already enabled")
+	}
+
+	adapter.outputTracker = make(map[string]struct{})
+}
+
+func (adapter *adapter) CleanupOutputs(ctx context.Context, outputs ...resource.Kind) error {
+	if adapter.outputTracker == nil {
+		panic("output tracking not enabled")
+	}
+
+	for _, outputKind := range outputs {
+		list, err := adapter.List(ctx, outputKind)
+		if err != nil {
+			return fmt.Errorf("error listing output resources: %w", err)
+		}
+
+		for _, resource := range list.Items {
+			if resource.Metadata().Owner() != adapter.name {
+				// skip resources not owned by this controller
+				continue
+			}
+
+			trackingID := outputTrackingID(resource.Metadata())
+
+			if _, touched := adapter.outputTracker[trackingID]; touched {
+				// skip touched resources
+				continue
+			}
+
+			if err = adapter.Destroy(ctx, resource.Metadata()); err != nil {
+				return fmt.Errorf("error destroying resource %s: %w", resource.Metadata(), err)
+			}
+		}
+	}
+
+	adapter.outputTracker = nil
```

Comment:
We can preserve map capacity here:
```suggestion
	for k := range adapter.outputTracker {
		delete(adapter.outputTracker, k)
	}
```
This will be compiled `map_clear` I think.

Reply:
Or! Add global pool and use it. That way we can preserve check in `StartTrackingOutputs`

Reply:
no, I need explicit `nil` to make sure the calls are always properly paired: `StartTrackingOutputs()` & `CleanupOutputs()`, `nil` is an explicit signal here.

Reply:
added a sync.Pool, I somehow didn't see your sync.Pool comment first :-D